src/apps/competitions/unpackers/base_unpacker.py

```diff
@@ -0,0 +1,294 @@
+import base64
+import json
+import os
+import uuid
+
+from django.core.files import File
+from django.utils import timezone
+
+from api.serializers.competitions import CompetitionSerializer
+from api.serializers.tasks import TaskSerializer, SolutionSerializer
+from competitions.models import Phase
+from datasets.models import Data
+from tasks.models import Task, Solution
+from .utils import CompetitionUnpackingException, zip_if_directory
+
+
+class BaseUnpacker:
+    def __init__(self, competition_yaml, temp_directory, creator):
+        self.competition_yaml = competition_yaml
+        self.temp_directory = temp_directory
+        self.creator = creator
+        self.competition = None
+        self.created_tasks = []
+        self.created_solutions = []
+        self.created_datasets = []
+
+    def _get_data_key(self, file_name, file_path, file_type, creator, *args, **kwargs):
+        if os.path.exists(file_path):
+            new_dataset = Data(
+                created_by_id=creator,
+                type=file_type,
+                name=f"{file_type} @ {timezone.now():'%m-%d-%Y %H:%M'}",
+                was_created_by_competition=True,
+            )
+            file_path = zip_if_directory(file_path)
+            new_dataset.data_file.save(os.path.basename(file_path), File(open(file_path, 'rb')))
+            self.created_datasets.append(new_dataset)
+            return new_dataset.key
+        else:
+            try:
+                # checking if file name could be a uuid
+                uuid.UUID(file_name)
+            except ValueError:
+                raise CompetitionUnpackingException(f'Cannot find dataset: "{file_name}"')
+
+            if not Data.objects.filter(key=file_name).exists():
+                raise CompetitionUnpackingException(f'Cannot find {file_type} with key: "{file_name}"')
+            return file_name
+
+    def _read_image(self, image_name):
+        image_path = os.path.join(self.temp_directory, image_name)
+        try:
+            with open(image_path, "rb") as image:
+                return json.dumps({
+                    "file_name": image_name,
+                    # Converts to b64 then to string
+                    "data": base64.b64encode(image.read()).decode()
+                })
+        except FileNotFoundError:
+            raise CompetitionUnpackingException(f"Unable to find image: {self.competition_yaml.get('image')}")
+
+    def _get_current_phase(self, phases):
+        for phase in phases:
+            if phase['start'] < timezone.now():
+                try:
```

Comment:
May be prettier like this:

```
if phase['start'] < timezone.now():
    no_end = 'end' not in phase or phase['end'] is None
    ends_in_future = phase['end'] > timezone.now()
    if no_end or ends_in_future:
        return phase
```

idk, is that nicer to you? not a huge deal at all

Reply:
yea. `ends_in_future` will raise a key error though. buut we could do:
```python
if phase['start'] < timezone.now():
    no_end = 'end' not in phase or phase['end'] is None
    if no_end or phase['end'] > timezone.now():
        return phase
```
This should prevent a KeyError from getting raised, right?